include/core/CCompressedLfuCache.h

```diff
@@ -0,0 +1,760 @@
+            return m_Count < rhs.count() ||
+                   (m_Count == rhs.count() && m_MemoryUsage > rhs.memoryUsage());
+        }
+
+        TCompressedKey key() const { return m_Key; }
+        std::size_t memoryUsage() const { return m_MemoryUsage; }
+        std::size_t count() const { return m_Count; }
+        void count(std::size_t count) { m_Count = count; }
+
+        void acceptPersistInserter(CStatePersistInserter& inserter) const {
+            CPersistUtils::persist(COUNT_TAG, m_Count, inserter);
+            CPersistUtils::persist(KEY_TAG, m_Key, inserter);
+            CPersistUtils::persist(MEMORY_USAGE_TAG, m_MemoryUsage, inserter);
+        }
+
+        bool acceptRestoreTraverser(CStateRestoreTraverser& traverser) {
+            do {
+                const std::string& name{traverser.name()};
+                RESTORE_WITH_UTILS(COUNT_TAG, m_Count)
+                RESTORE_WITH_UTILS(KEY_TAG, m_Key)
+                RESTORE_WITH_UTILS(MEMORY_USAGE_TAG, m_MemoryUsage)
+            } while (traverser.next());
+            return true;
+        }
+
+    private:
+        static const std::string COUNT_TAG;
+        static const std::string KEY_TAG;
+        static const std::string MEMORY_USAGE_TAG;
+
+    private:
+        TCompressedKey m_Key;
+        std::size_t m_MemoryUsage{0};
+        std::uint64_t m_Count{0};
+    };
+
+    using TCacheItemStatsSet = std::multiset<CCacheItemStats>;
+    using TCacheItemStatsSetItr = typename TCacheItemStatsSet::iterator;
+
+    //! \brief A cache item wrapper which holds a reference to its stats.
+    class CCacheItem {
+    public:
+        //! We purposely disable direct memory estimation for cache items because
+        //! we want to avoid quadratic complexity using core::CMemory::dynamicSize.
+        //! Instead we estimate memory usage for each item we add and remove from
+        //! the cache and maintain a running total.
+        static bool dynamicSizeAlwaysZero() { return true; }
+
+    public:
+        CCacheItem() = default;
+        explicit CCacheItem(VALUE value) : m_Value{std::move(value)} {}
+
+        const VALUE& value() const { return m_Value; }
+        TCacheItemStatsSetItr stats() const { return m_Stats; }
+        void stats(TCacheItemStatsSetItr stats) { m_Stats = stats; }
+
+        void acceptPersistInserter(CStatePersistInserter& inserter) const {
+            CPersistUtils::persist(VALUE_TAG, m_Value, inserter);
+        }
+
+        bool acceptRestoreTraverser(CStateRestoreTraverser& traverser) {
+            do {
+                const std::string& name{traverser.name()};
+                RESTORE_WITH_UTILS(VALUE_TAG, m_Value)
+            } while (traverser.next());
+            return true;
+        }
+
+    private:
+        static const std::string VALUE_TAG;
+
+    private:
+        VALUE m_Value;
+        TCacheItemStatsSetItr m_Stats;
+    };
+
+    //! \brief An atomic 128 bit counter.
+    class CAtomicLargeCounter {
```

Comment:
I think there is an element of over engineering using a 128bit counter for the hit count.

If this class is useful and can be reused shouldn't it been in a dedicated file?

Reply:
Yes, I agree.  A 64 bit counter won't wrap for over 67 years if it's incremented 4 billion times per second.  So in reality `m_Upper` is always going to be zero and this class is just adding complexity to the codebase.

Reply:
Ok happy to remove. I didn't feel 100% comfortable since we don't actually know how this cache is used: it could at some point be used for something which receives a requests, but I guess we just ignore for now. It is only for monitoring anyway.

Reply:
See [375e1fd](https://github.com/elastic/ml-cpp/pull/2305/commits/375e1fdec34367592a3714223b28c993a3541eeb).